tools/hasher/src/main/java/org/apache/shiro/tools/hasher/Hasher.java

```diff
@@ -495,6 +495,9 @@ private static char[] readPassword(boolean confirm) throws IOException {
             throw new IllegalArgumentException("No password specified.");
         }
         if (confirm) {
+            if (!isTerminal(console)) {
+                throw new IllegalStateException("Cannot confirm password when console is not available.");
+            }
```

Comment:
Maybe on non-Terminals, just  use one input instead of to (i.e.  do not throw an exception and just go for it).
Otherwise this might break script invocations.

Reply:
great idea, done